pestion/src/lib.rs

```diff
@@ -0,0 +1,443 @@
+//!     assert_eq!("atomic", a_rule.get("type").and_then(|t| t.as_str()).unwrap());
+//!
+//!     // the first node in the expression tree is a `choice` operator
+//!     assert_eq!(
+//!         "choice",
+//!         a_rule
+//!             .get("expression")
+//!             .and_then(|e| e.as_sequence())
+//!             .and_then(|s| s.get(0))
+//!             .and_then(|h| h.as_str()).unwrap()
+//!     );
+//!     
+//!     Ok(())
+//! }
+//! ```
+//!
+//! [Pest]: https://pest.rs/
+//! [Ion]: https://amzn.github.io/ion-docs/
+
+pub mod result;
+
+pub use result::*;
+
+use ion_rs::value::owned::{text_token, OwnedElement, OwnedValue};
+use ion_rs::value::{Builder, Element};
+use pest::Parser;
+use pest_meta::ast::{Expr, Rule as AstRule, RuleType as AstRuleType, RuleType};
+use pest_meta::parser::{consume_rules, PestParser, Rule};
+
+/// Converts a representation of a Pest grammar (or part of a grammar) into Ion [`Element`].
+pub trait PestToElement {
+    type Element: Element;
+
+    /// Converts this into [`Element`] which may imply parsing Pest syntax.
+    fn pest_to_element(&self) -> PestionResult<Self::Element>;
+}
+
+impl PestToElement for &str {
+    type Element = OwnedElement;
+
+    /// Parses a `str` slice as a Pest grammar and serializes the AST into [`Element`].
+    fn pest_to_element(&self) -> PestionResult<Self::Element> {
+        let pairs = PestParser::parse(Rule::grammar_rules, *self)?;
+        let ast = match consume_rules(pairs) {
+            Ok(ast) => ast,
+            Err(errors) => {
+                return if errors.is_empty() {
+                    invalid("Error converting Pest grammar to AST with no context")
+                } else {
+                    // TODO deal with more than one error..
+                    let err = errors.into_iter().next().unwrap();
+                    Err(err.into())
+                };
+            }
+        };
+
+        ast.pest_to_element()
+    }
+}
+
+impl PestToElement for Vec<AstRule> {
+    type Element = OwnedElement;
+
+    /// Converts a body of rules into a `struct` that has a rule for each field.
+    fn pest_to_element(&self) -> PestionResult<Self::Element> {
+        let mut fields = vec![];
+        for rule in self.iter() {
+            let rule_name = text_token(rule.name.clone());
+            let rule_value = rule.pest_to_element()?;
+            fields.push((rule_name, rule_value));
+        }
+        Ok(Self::Element::new_struct(fields))
+    }
+}
+
+impl PestToElement for AstRule {
+    type Element = OwnedElement;
+
+    /// Converts a Pest Rule into a `struct` that has the field for [`RuleType`] as a symbol
+    /// and a field for the [`Expr`].
+    fn pest_to_element(&self) -> PestionResult<Self::Element> {
+        let fields = vec![
+            (text_token("type"), self.ty.pest_to_element()?),
+            (text_token("expression"), self.expr.pest_to_element()?),
+        ];
+        Ok(Self::Element::new_struct(fields))
```

Comment:
[`Element::new_struct`](https://github.com/amzn/ion-rust/blob/d1768e8f18a11695181a16e379e426780d865b39/src/value/mod.rs#L707-L712) takes an iterator:

```rust
    /// Builds a `struct` from Struct using Builder.
    fn new_struct<K, V, I>(structure: I) -> Self::Element
    where
        K: Into<Self::SymbolToken>,
        V: Into<Self::Element>,
        I: IntoIterator<Item = (K, V)>;
```

You could make `fields` into a slice to save yourself an allocation.

Reply:
I could not make it into a slice because I need ownership of the tuple that we construct the `struct` with, However, I am able to write it as:

```rust
        let fields = std::array::IntoIter::new([
            (text_token("type"), self.ty.pest_to_element()?),
            (text_token("expression"), self.expr.pest_to_element()?),
        ]);
        Ok(Self::Element::new_struct(fields))
```

Note that the explicit `IntoIter::new` is needed because sadly, arrays don't implement `IntoIter`...

Reply:
Updated in [783de7f](https://github.com/partiql/partiql-lang-rust/pull/38/commits/783de7fae7c9f175f33af6e661a8c843cc49037a)

Reply:
Thanks! The explicit `IntoIter::new` is a bit unfortunate, but happily it looks like [arrays will get a stable implementation of `IntoIterator` in the next Rust release](https://github.com/rust-lang/rust/pull/65819#issuecomment-826859927) (v1.53).